yuyo/modals.py

```diff
@@ -0,0 +1,1440 @@
+
+
+_ModalResponseT = typing.Union[hikari.api.InteractionMessageBuilder, hikari.api.InteractionDeferredBuilder]
+"""Type hint of the builder response types allows for modal interactions."""
+
+
+class _NoDefaultEnum(enum.Enum):
+    VALUE = object()
+
+
+NO_DEFAULT = _NoDefaultEnum.VALUE
+"""Singleton used to signify when a field has no default."""
+
+NoDefault = typing.Literal[_NoDefaultEnum.VALUE]
+"""Type of [yuyo.modals.NO_DEFAULT][]."""
+
+
+class AbstractTimeout(abc.ABC):
+    """Abstract interface used to manage timing out a modal."""
+
+    __slots__ = ()
+
+    @property
+    @abc.abstractmethod
+    def has_expired(self) -> bool:
+        """Whether this modal has timed-out."""
+
+    @abc.abstractmethod
+    def increment_uses(self) -> bool:
+        """Add a use to the modal.
+
+        Returns
+        -------
+        bool
+            Whether the modal has now timed-out.
+        """
+
+
+class BasicTimeout(AbstractTimeout):
+    """Basic modal timeout strategy.
+
+    This implementation timeouts if `timeout` passes since the last call or
+    when `max_uses` reaches `0`.
+    """
+
+    __slots__ = ("_last_triggered", "_timeout", "_uses_left")
+
+    def __init__(self, timeout: typing.Union[datetime.timedelta, int, float], /, *, max_uses: int = -1) -> None:
+        """Initialise a basic timeout.
+
+        Parameters
+        ----------
+        timeout
+            How long this modal should wait between calls before timing-out.
+        max_uses
+            The maximum amount of uses this modal allows.
+
+            Setting this to `-1` marks it as unlimited.
+        """
+        if not isinstance(timeout, datetime.timedelta):
+            timeout = datetime.timedelta(seconds=timeout)
+
+        self._last_triggered = datetime.datetime.now(tz=datetime.timezone.utc)
+        self._timeout = timeout
+        self._uses_left = max_uses
+
+    @property
+    def has_expired(self) -> bool:
+        # <<inherited docstring from AbstractTimeout>>.
+        if self._uses_left == 0:
+            return True
+
+        return datetime.datetime.now(tz=datetime.timezone.utc) - self._last_triggered > self._timeout
+
+    def increment_uses(self) -> bool:
+        # <<inherited docstring from AbstractTimeout>>.
+        if self._uses_left > 0:
+            self._uses_left -= 1
+
+        elif self._uses_left == 0:
+            raise RuntimeError("Uses already depleted")
+
+        self._last_triggered = datetime.datetime.now(tz=datetime.timezone.utc)
+        return self._uses_left == 0
+
+
+class NeverTimeout(AbstractTimeout):
+    """Timeout implementation which never expires."""
+
+    __slots__ = ()
+
+    @property
+    def has_expired(self) -> bool:
+        # <<inherited docstring from AbstractTimeout>>.
+        return False
+
+    def increment_uses(self) -> bool:
+        # <<inherited docstring from AbstractTimeout>>.
+        return False
+
+
+class ModalContext(components_.BaseContext[hikari.ModalInteraction]):
```

Comment:
No dataclasses? 
![image](https://user-images.githubusercontent.com/11181703/220190257-ddfc0de9-00d3-40b3-a95d-3b60c80edfb8.png)

Reply:
Meant to comment this on Modal but shh

Reply:
Maybe another time, not having alias is annoying for the internal fields